core/src/main/java/org/apache/iceberg/util/Tasks.java

```diff
@@ -385,128 +401,117 @@ public void run() {
+            matchedRetryException = true;
+          }
+        }
+        if (!matchedRetryException) {
+          return true;
+        }
 
-          try {
-            TimeUnit.MILLISECONDS.sleep(delayMs + jitter);
-          } catch (InterruptedException ie) {
-            Thread.currentThread().interrupt();
-            throw new RuntimeException(ie);
+      } else {
+        // otherwise, always retry unless one of the stop exceptions is found
+        for (Class<? extends Exception> exClass : stopRetryExceptions) {
+          if (exClass.isInstance(exception)) {
+            return true;
           }
         }
       }
+      return false;
     }
   }
 
-  private static Collection<Throwable> waitFor(Collection<Future<?>> futures)
-      throws Error {
+  private static Collection<Throwable> waitFor(Collection<Future<?>> futures) {
     while (true) {
-      int numFinished = 0;
-      for (Future<?> future : futures) {
-        if (future.isDone()) {
-          numFinished += 1;
-        }
-      }
-
+      long numFinished = futures.stream().filter(Future::isDone).count();
       if (numFinished == futures.size()) {
         List<Throwable> uncaught = new ArrayList<>();
-          // all of the futures are done, get any uncaught exceptions
-          for (Future<?> future : futures) {
-            try {
-              future.get();
-
-            } catch (InterruptedException e) {
-              LOG.warn("Interrupted while getting future results", e);
-              for (Throwable t : uncaught) {
-                e.addSuppressed(t);
-              }
-              Thread.currentThread().interrupt();
-              throw new RuntimeException(e);
-
-            } catch (CancellationException e) {
-              // ignore cancellations
-
-            } catch (ExecutionException e) {
-              Throwable cause = e.getCause();
-              if (Error.class.isInstance(cause)) {
-                for (Throwable t : uncaught) {
-                  cause.addSuppressed(t);
-                }
-                throw (Error) cause;
-              }
-
-              if (cause != null) {
-                uncaught.add(e);
-              }
-
-              LOG.warn("Task threw uncaught exception", cause);
-            }
-          }
-
+        // all of the futures are done, get any uncaught exceptions
+        futures.forEach(future -> uncaught.addAll(extractUncaughtException(future)));
         return uncaught;
-
       } else {
         try {
           Thread.sleep(10);
         } catch (InterruptedException e) {
           LOG.warn("Interrupted while waiting for tasks to finish", e);
-
-          for (Future<?> future : futures) {
-            future.cancel(true);
-          }
+          futures.forEach(future -> future.cancel(true));
           Thread.currentThread().interrupt();
           throw new RuntimeException(e);
         }
       }
     }
   }
 
+  private static List<Throwable> extractUncaughtException(Future<?> future) {
```

Comment:
Nit: returns a list so the method name should be plural.

Reply:
Nevermind, this is for a single future. It should only return a single exception then.